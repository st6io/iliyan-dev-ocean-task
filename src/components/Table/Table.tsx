import styled, { SystemProps, x } from '@xstyled/styled-components';

interface Props extends SystemProps {
  rows: RowType[];
  headers?: string[];
  variant?: VariantType;
  onRowClick?: (row: RowType) => void;
}

export enum Variant {
  Primary = 'primary',
  Secondary = 'secondary',
}
export type VariantType = `${Variant}`;

type RowType = {
  id: string;
  cells: string[];
};

const backgroundByVariant = {
  [Variant.Primary]: 'white',
  [Variant.Secondary]: 'light-grey',
};

const Cell = styled(x.td)`
  text-align: start;
  padding: 5;
`;

const StyledTable = styled(x.table)`
  // xstyled doesn't support utility prop "borderSpacing"
  border-spacing: 0 0.4rem;
`;

const Table = ({
  rows,
  onRowClick,
  headers = [],
  variant = Variant.Primary,
  ...otherProps
}: Props) => (
  <StyledTable {...otherProps}>
    {headers.length ? (
      <x.thead position="sticky" top={0}>
        <x.tr bg={backgroundByVariant[variant]}>
          {headers.map((header) => (
            <Cell as="th" key={header} textTransform="uppercase" color="primary">
              {header}
            </Cell>
          ))}
        </x.tr>
      </x.thead>
    ) : null}

    <x.tbody>
      {rows.map((row) => (
        <x.tr
          key={row.id}
          bg={{ _: backgroundByVariant[variant], hover: onRowClick ? 'light-grey' : null }}
          cursor={onRowClick ? 'pointer' : 'auto'}
          onClick={() => onRowClick?.(row)}
        >
          {row.cells.map((cell) => (
            <Cell key={cell} color={`text.${variant}`}>
              {cell}
            </Cell>
          ))}
        </x.tr>
      ))}
    </x.tbody>
  </StyledTable>
);

export default Table;
