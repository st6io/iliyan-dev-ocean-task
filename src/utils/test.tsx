import { ReactElement, ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { render as nativeRender } from '@testing-library/react';

const render = (ui: ReactElement, { path = '/', ...options } = {}) => {
  const RouterWrapper = ({ children }: { children: ReactNode }) => (
    <MemoryRouter initialEntries={[path]}>{children}</MemoryRouter>
  );

  return nativeRender(ui, {
    wrapper: RouterWrapper,
    ...options,
  });
};

export * from '@testing-library/react';
export { render };
