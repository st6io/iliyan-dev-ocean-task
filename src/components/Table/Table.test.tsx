import { render, screen } from '@src/utils/test';

import Table from './Table';

const headers = ['name', 'description'];
const rows = new Array(5).fill(null).map((_, index) => ({
  id: `id-${index}`,
  cells: [`name ${index}`, `description ${index}`],
}));

const renderTable = (props = {}) => ({
  ...render(<Table headers={headers} rows={rows} {...props} />),
  queryThead: () => screen.queryByRole('table')?.querySelector('thead'),
  queryTbody: () => screen.queryByRole('table')?.querySelector('tbody'),
});

describe('Table', () => {
  it('should match snapshot', () => {
    const { container } = renderTable();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should respect otherProps', () => {
    renderTable({ w: 0.1234 });

    expect(screen.queryByRole('table')).toHaveStyle({
      width: '12.34%',
    });
  });

  it('should apply correct styles', () => {
    renderTable();

    expect(screen.queryByRole('table')).toHaveStyle({
      'border-spacing': '0 0.4rem',
    });
  });

  describe('head', () => {
    it('should have a sticky thead row', () => {
      const { queryThead } = renderTable();

      expect(queryThead()).toHaveStyle({
        position: 'sticky',
        top: 0,
      });
    });

    it('should render header cells with correct styles', () => {
      renderTable({ headers });

      headers.forEach((header) => {
        const headerCell = screen.queryByText(header);
        expect(headerCell).toBeInTheDocument();
        expect(headerCell).toHaveStyle({
          padding: '1.25rem',
          'text-align': 'start',
          color: '#b41d76',
          'text-transform': 'uppercase',
        });
      });
    });

    it('should NOT render thead row when no headers', () => {
      const { queryThead } = renderTable({ headers: undefined });

      expect(queryThead()).not.toBeInTheDocument();
    });
  });

  describe('body', () => {
    it('should render cells with correct styles', () => {
      renderTable({ rows });

      rows
        .flatMap(({ cells }) => cells)
        .forEach((cell) => {
          const cellElement = screen.queryByText(cell);
          expect(cellElement).toBeInTheDocument();
          expect(cellElement).toHaveStyle({
            padding: '1.25rem',
            'text-align': 'start',
            color: '#353449',
          });
        });
    });
  });

  describe('variants', () => {
    it.each([
      {
        testName: 'should apply "primary" variant styles by default',
        variant: undefined,
        expected: {
          bg: 'rgb(255, 255, 255)',
          color: 'rgb(53, 52, 73)',
        },
      },
      {
        testName: 'should apply "secondary" variant styles',
        variant: 'secondary',
        expected: {
          bg: 'rgb(248, 248, 250)',
          color: 'rgb(146, 146, 157)',
        },
      },
    ])('$testName', ({ variant, expected }) => {
      renderTable({ rows, variant });

      const trElements = screen.queryByRole('table')?.querySelectorAll('tr') || [];

      expect(trElements.length).toBeGreaterThan(0);
      trElements.forEach((rowElement) => {
        expect(rowElement).toHaveStyle({
          'background-color': expected.bg,
        });
      });

      rows
        .flatMap(({ cells }) => cells)
        .forEach((cell) => {
          const cellElement = screen.queryByText(cell);
          expect(cellElement).toBeInTheDocument();
          expect(cellElement).toHaveStyle({
            color: expected.color,
          });
        });
    });
  });
});
