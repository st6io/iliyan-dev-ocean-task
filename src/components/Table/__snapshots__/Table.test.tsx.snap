// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match snapshot 1`] = `
<table
  class="sc-WCkqM sc-grVGCS  eKfFBe"
>
  <thead
    class="sc-kYWVYA kzCKnj"
  >
    <tr
      class="sc-jGprRt dezIds"
    >
      <th
        class="sc-tsFYE sc-iJkHyd jgyunm JjpSb"
      >
        name
      </th>
      <th
        class="sc-tsFYE sc-iJkHyd jgyunm JjpSb"
      >
        description
      </th>
    </tr>
  </thead>
  <tbody
    class="sc-ikjQzJ"
  >
    <tr
      class="sc-jGprRt dBZKRU"
    >
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        name 0
      </td>
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        description 0
      </td>
    </tr>
    <tr
      class="sc-jGprRt dBZKRU"
    >
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        name 1
      </td>
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        description 1
      </td>
    </tr>
    <tr
      class="sc-jGprRt dBZKRU"
    >
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        name 2
      </td>
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        description 2
      </td>
    </tr>
    <tr
      class="sc-jGprRt dBZKRU"
    >
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        name 3
      </td>
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        description 3
      </td>
    </tr>
    <tr
      class="sc-jGprRt dBZKRU"
    >
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        name 4
      </td>
      <td
        class="sc-tsFYE sc-iJkHyd liZNfH JjpSb"
      >
        description 4
      </td>
    </tr>
  </tbody>
</table>
`;
